test/common/testlib.py

```diff
@@ -745,27 +765,34 @@ def setUp(self):
         if opts.address and self.provision is not None:
             raise unittest.SkipTest("Cannot provision multiple machines if a specific machine address is specified")
 
-        self.machine = None
         self.browser = None
         self.machines = {}
         provision = self.provision or {'machine1': {}}
 
-        # First create all machines, wait for them later
-        for key in sorted(provision.keys()):
-            options = provision[key].copy()
-            if 'address' in options:
-                del options['address']
-            if 'dns' in options:
-                del options['dns']
-            if 'dhcp' in options:
-                del options['dhcp']
-            machine = self.new_machine(**options)
-            self.machines[key] = machine
-            if not self.machine:
-                self.machine = machine
-            if opts.trace:
-                print("Starting {0} {1}".format(key, machine.label))
-            machine.start()
+        test_method = getattr(self.__class__, self._testMethodName)
+        if getattr(test_method, "_testlib__non_destructive", False) and not opts.address:
+            if self.provision:
+                raise unittest.SkipTest("Cannot provision machines if test is marked as nondestructive")
+            self.machines['machine1'] = MachineCase.get_global_machine()
+            self.machine = MachineCase.get_global_machine()
```

Comment:
Style nitpick: It should be idempotent, but it's both clearer to read as well as faster to re-use the variable. I. e.

    self.machines['machine1'] = self.machine = MachineCase.get_global_machine()